storage/innobase/fil/fil0crypt.cc

```diff
@@ -635,36 +694,112 @@ static fil_space_crypt_t *fil_space_set_crypt_data(
-  ut_ad((memcmp(ptr, ENCRYPTION_KEY_MAGIC_PS_V1, ENCRYPTION_MAGIC_SIZE) == 0));
+  ut_ad((memcmp(ptr, ENCRYPTION_KEY_MAGIC_PS_V2, ENCRYPTION_MAGIC_SIZE) == 0));
   ptr += ENCRYPTION_MAGIC_SIZE;
 
-  ptr += 2;  // length of iv has been already read
+  uint type = mach_read_from_1(ptr);
+  ptr += 1;
+
+  ut_a(type == CRYPT_SCHEME_UNENCRYPTED ||
+       type == CRYPT_SCHEME_1);  // only supported
 
-  space_id_t space_id = mach_read_from_4(ptr);
+  uint min_key_version = mach_read_from_4(ptr);
   ptr += 4;
-  uint offset = mach_read_from_2(ptr);
-  ptr += 2;
+
+  uint key_id = mach_read_from_4(ptr);
+  ptr += 4;
+
+  fil_encryption_t encryption = (fil_encryption_t)mach_read_from_1(ptr);
+  ptr += 1;
+
+  fil_space_crypt_t *crypt_data = fil_space_create_crypt_data(
+      encryption, key_id, Crypt_key_operation::FETCH_OR_GENERATE_KEY);
+  /* Need to overwrite these as above will initialize fields. */
+  DBUG_ASSERT(min_key_version != ENCRYPTION_KEY_VERSION_INVALID);
+  crypt_data->min_key_version = min_key_version;
+  crypt_data->encryption = encryption;
+  memcpy(crypt_data->iv, ptr, CRYPT_SCHEME_1_IV_LEN);
+  ptr += CRYPT_SCHEME_1_IV_LEN;
+  crypt_data->encryption_rotation =
+      static_cast<Encryption_rotation>(mach_read_from_1(ptr));
+  ptr += 1;
+  uchar tablespace_key[ENCRYPTION_KEY_LEN];
+  memcpy(tablespace_key, ptr, ENCRYPTION_KEY_LEN);
+  ptr += ENCRYPTION_KEY_LEN;
+
+  if (std::search_n(tablespace_key, tablespace_key + ENCRYPTION_KEY_LEN,
+                    ENCRYPTION_KEY_LEN,
+                    0) ==
+      tablespace_key) {  // tablespace_key is all zeroes which means there is no
+                         // tablepsace in mtr log
+    crypt_data->set_tablespace_key(nullptr);
+    ptr += ENCRYPTION_KEY_LEN;
+  } else {
+    crypt_data->set_tablespace_key(tablespace_key);
+  }
+
+  /* Check is used key found from encryption plugin */
+  if (crypt_data->should_encrypt() && !crypt_data->is_key_found()) {
+    ib::error() << "Key cannot be read for space id = " << space_id;
+    recv_sys->set_corrupt_log();
+  }
+
+  /* update fil_space memory cache with crypt_data */
+  if (fil_space_t *space = fil_space_acquire_silent(space_id)) {
+    crypt_data = fil_space_set_crypt_data(space, crypt_data);
+    fil_space_release(space);
+  } else {
+    fil_space_destroy_crypt_data(&crypt_data);
+  }
+
+  return ptr;
+}
+/******************************************************************
+Parse a MLOG_FILE_WRITE_CRYPT_DATA log entry
+@param[in]  space_id  id of space that this log entry refers to
+@param[in]  ptr  Log entry start
+@param[in]  end_ptr  Log entry end
+@param[in]  len  Log entry length
+@return position on log buffer */
+byte *fil_parse_write_crypt_data_v1(space_id_t space_id, byte *ptr,
+                                    const byte *end_ptr, ulint len) {
+  ptr += 4;  // skip offset and len
+  ptr += 2;  // skip iv_length
+
+  if (len != KERYING_ENCRYPTION_INFO_MAX_SIZE) {
```

Comment:
How can this work?

In v1 KERYING_ENCRYPTION_INFO_MAX_SIZE should be different and
v2 KERYING_ENCRYPTION_INFO_MAX_SIZE size is something else.

Ideally, we should have KERYING_ENCRYPTION_INFO_MAX_SIZE_57 with old data layout and old fields(struct).

Reply:
right , added KEYRING_ENCRYPTION_INFO_MAX_SIZE_V1 - to keep the naming inline with crypt_data_v1 and v2.